### ISSUE TITLE
Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32

### ISSUE THREAD

**Issue by carlosgalvezp** [onnx/tensorflow-onnx#883]
Hi,

I'm trying to use the TensorFlow-ONNX-TensorRT workflow described here:

https://devblogs.nvidia.com/speeding-up-deep-learning-inference-using-tensorflow-onnx-and-tensorrt/

Versions:
Tensorflow: 1.14.0 (GPU)
ONNX: 1.6.0
Tensorflow-onnx: master at the time of writing
TensorRT: 7.0

I could convert my TensorFlow model to ONNX successfully with opset 11. However when passing it through "trtexec" I get the following errors:

```
[V] [TRT] ModelImporter.cpp:90: Importing initializer: const_fold_opt__961
[W] [TRT] onnx2trt_utils.cpp:198: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
```

Why are the weights not FP32? In some cases it even fails to convert to INT32:

```
BatchToSpaceND_int_max__752__753
[W] [TRT] onnx2trt_utils.cpp:198: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[V] [TRT] onnx2trt_utils.cpp:212: Weight at index 0: 9223372036854775807 is out of range. Clamping to: 2147483647
```

It seems the errors come from layers inserted by the ONNX converter. 

Thanks!

**Comment by jignparm** [comment]
>Your ONNX model has been generated with INT64 weights

In Onnx, some operators require `int_max` or `int_min` as special values to denote 'infinity'(e.g. Slice operator), which is probably where the large integer values are coming from.

TRT seems to be doing the right thing above by clamping and/or casting to INT32 however. Are these treated as critical errors, or just warnings by TRT?

**Comment by carlosgalvezp** [comment]
Hi!

It's indeed "just" a warning, but it's not a good idea to keep warnings, they produce noise and distract the user from real warnings and errors that should be taken care off. Why does ONNX need 64-bit integers for these parameters? Infinity can very well be represented with 32-bit integers? I think fixing this warning would greatly improve the user experience :)

Anyway the actual crash I got was here:

```
[V] [TRT] ModelImporter.cpp:107: Parsing node: Main_conv_dilated/SpaceToBatchND_pad__833 [Pad]
[V][TRT] ModelImporter.cpp:123: Searching for input: const_fold_opt__939
[V][TRT] ModelImporter.cpp:123: Searching for input: Main_conv_dilated/SpaceToBatchND__filltop_const__705__706
[V] [TRT] ModelImporter.cpp:129: Main_conv_dilated/SpaceToBatchND_pad__833 [Pad] inputs: [const_fold_opt__939 -> (3, 2)], [Main_conv_dilated/SpaceToBatchND__filltop_const__705__706 -> (4)], 
terminate called after throwing an instance of 'std::out_of_range'
  what():  Attribute not found: pads
Aborted (core dumped)

```

Let me know if you'd like me to open a separate issue for this!

**Comment by jignparm** [comment]
@carlosgalvezp, a few questions below regarding TensorRT.

   * Does TensorRT only support 32-BIT ONNX operators?
      * If **YES**, then how are 64-BIT  ops like `Pad` and `Shape` supported?
         * For example, the `pads` attribute of the `Pad` operator only supports INT64 type values. From the exception above, it seems like TRT is complaining about this attribute. 
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Pad
         * Another example is the `Shape` op, where is output is always INT64.
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Shape
      * If **NO**, is there another doc link besides this one?
https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/python_api/uff/Operators.html
 
TF2ONNX generates models using the full suite of ONNX operators, but since TRT only supports 32-BIT values, it may need to resort to casting or clamping to convert a full ONNX model into a TRT-ONNX model (just a hypothesis).

Tf2onnx composes the SpaceToBatchND op using a collection of smaller operators like Pad and Shape, so there's no way to completely avoid INT64 type. The majority of ONNX models would need to use INT64 somewhere, because of the primitive ops above -- so it's likely that TRT is doing some downcasting to get other models to work as well, but failing for this particular model somehow.

**Comment by carlosgalvezp** [comment]
@jignparm Thanks for the reply. Here's what I can find:

- TensorRT does not seem to support INT64 whatsoever:
https://docs.nvidia.com/deeplearning/sdk/tensorrt-support-matrix/index.html#layers-precision-matrix
In fact, from that table it seems like `IPaddingLayer` doesn't even support INT32. That seems to contradict the ONNX supported ops so I don't know what to make out of it:
https://docs.nvidia.com/deeplearning/sdk/tensorrt-support-matrix/index.html#supported-ops

- ONNX-TensorRT parser has some clarifications:
https://github.com/onnx/onnx-tensorrt/blob/master/operators.md
```
TensorRT supports the following ONNX data types: FLOAT32, FLOAT16, INT8, and BOOL

*There is limited support for INT32 and INT64 types. TensorRT will attempt to cast down INT64 to INT32 where possible. If not possible, TensorRT will throw an error. See the TensorRT layer support matrix for more information on data type support.
```
Also:
```
Pad | Y | Zero-padding on last 2 dimensions only
```

As you say, it seems TensorRT should be fine with simply downcasting to INT32. I don't understand why it's missing the "pads" attribute. Is there a good way to verify that the attribute is indeed in the .onnx file?

Btw about the second link that you posted, UFF will soon be deprecated so it's no longer relevant.

**Comment by carlosgalvezp** [comment]
Ok, apparently it has been solved very recently in latest master of ONNX-TRT, which I don't seem to be using (Tagged version 7.0 instead):
https://github.com/onnx/onnx-tensorrt/pull/408/commits

The reason seems to be that the padding information was changed between opset 10 and 11, and ONNX-TRT didn't catch up with it yet.

I'd gladly use opset 10 but somehow after the fix to my other issue https://github.com/onnx/tensorflow-onnx/issues/801 I need opset 11. Is it expected to require opset 11 for simple operations such as a Conv2D?

**Comment by jignparm** [comment]
>Is there a good way to verify that the attribute is indeed in the .onnx file?

1> One quick way is to run the model using OnnxRuntime. See this simple Python API  example here https://pypi.org/project/onnxruntime/0.5.0/  -- install the latest version 1.2.0. If the model runs successfully using OnnxRuntime, it means the the model is fine.

2> Use the Netron tool to visualize the model and click on the node to see the attribute https://github.com/lutzroeder/netron

**Comment by jignparm** [comment]
>Ok, apparently it has been solved very recently in latest master of ONNX-TRT, 

Great to see it's solved in latest TRT

> Is it expected to require opset 11 for simple operations such as a Conv2D?

In general, if there's even one operator in the TF model which requires opset 11, then we have to use opset 11 for the entire model -- there's no way to specify a opset version at the level of an individual operator.  So even though Conv2D is an early op, there may be other ops in the TF model with require a higher opset  --  using the highest opset  maximizes the chances of being able to convert the model successfully.

**Comment by carlosgalvezp** [comment]
> there may be other ops in the TF model with require a higher opset 
Good to know! Do you know where I can look which operators require which opset? Looking here https://github.com/onnx/tensorflow-onnx/blob/master/support_status.md I can only find up to opset 10. The requirement I got for opset 11 was coming from SpaceToBatchND. I'd gladly add more details but I feel I'm deviating a bit from the original bug.

Anyhow, I now tested latest ONNX-TRT and jumped into a new error :) 

```
While parsing node number 64 [Pad -> "a/Main_conv_dilated/SpaceToBatchND_pad__251:0"]:
--- Begin node ---
input: "const_fold_opt__912"
input: "b/Main_conv_dilated/SpaceToBatchND__filltop_const__787__788"
output: "a/Main_conv_dilated/SpaceToBatchND_pad__251:0"
name: "a/Main_conv_dilated/SpaceToBatchND_pad__251"
op_type: "Pad"
domain: ""

--- End node ---
ERROR: /opt/TensorRT/parsers/onnx/builtin_op_importers.cpp:2091 In function importPad:
[8] Assertion failed: tensor.getDimensions().nbDims >= 4

```

Why would the tensor have more dimensions than 4?

As I said this feels a bit outside of the "INT64" scope so do let me know if you want me to put this in a separate bug.

Thanks a log for the quick help!

**Comment by jignparm** [comment]
>Good to know! Do you know where I can look which operators require which opset? Looking here https://github.com/onnx/tensorflow-onnx/blob/master/support_status.md

Looks like the documentation here needs to be updated :) TF2onnx supports up to opset 11 now. Look for the documentation to be updated shortly.


> Why would the tensor have more dimensions than 4?

The specification of SpaceToBatchND does not limit the tensor dimensions to 4. Not sure if the assertion in TRT is a valid one.

Can you try running this model using OnnxRuntime? This will help to isolate between tf2onnx and TRT issues. If the model runs under OnnxRuntime, it means the model is valid, and tf2onnx should be OK. TRT may need to be updated to handle larger tensor values.

> if you want me to put this in a separate bug.

This thread should be good  -- looks like it's the errors are in the same vein.

**Comment by carlosgalvezp** [comment]
> Look for the documentation to be updated shortly

Awesome, will keep an eye! :)

> The specification of SpaceToBatchND does not limit the tensor dimensions to 4

Sure, but in our models we specify our inputs and outputs to be 4 dimensional, and nothing inside should change that dimensionality, right? I.e. why would a Conv2D create internally a tensor of >4 dimensions given a 4D input tensor?

> Can you try running this model using OnnxRuntime?

Will do asap.

> This thread should be good 

I was too impatient and created a new bug :(

**Comment by jignparm** [comment]
> and nothing inside should change that dimensionality, right?

Some of the composed operators may change dimensionality of the original tensor -- for example SpaceToBatchND requires projecting the original tensor into higher dimensions to manipulate the data blocks.

**Comment by VCBE123** [comment]
I can successfully convert pytorch-efficient-b2 model to onnx model and the outputs match with Pytorch model. 
But when I load the onnx model with TensorRT7.0 , it throws som pad warning and the outputs is not the same.

**Comment by jignparm** [comment]
@VCBE123  -- see the comments above. @carlosgalvezp saw  similar error  but fixed it by updating TensorRT, to solve the 'pads' issue.

**Comment by carlosgalvezp** [comment]
@jignparm 
> SpaceToBatchND requires projecting the original tensor into higher dimensions

If I understood you correctly, Conv2D is implemented in terms of SpaceToBatchND, which in turn needs to project the original tensor into higher dimensions. Then Conv2D should never have worked with TensorRT, right? Since TRT only can handle 4D tensors (for Conv2D).

I'm trying to understand what changed, Conv2D is a pretty basic operation and has presumably always worked with TRT for long time, right?

Perhaps the issue is related to the other bug that I opened:
https://github.com/onnx/tensorflow-onnx/issues/887

To support Conv2D('channels_first'), which only has GPU implementation, it feels like now extra operations are added to project into higher dimensions. Would that explain what I'm observing?

**Comment by jignparm** [comment]
> If I understood you correctly, Conv2D is implemented in terms of SpaceToBatchND,

I meant something slightly different. The `Conv2D` operator in TF is mapped directly to the `Conv` operator in ONNX, so that is not the problem.

From your log above, the entry `"a/Main_conv_dilated/SpaceToBatchND_pad__251:0"`  suggests the TF model contains a SpaceToBatchND op. There is no equivalent single ONNX op, and the only way to support it in tf2onnx is to compose it using a network/collection of  ONNX ops. In these operations, 4D tensors are reshaped temporarily into 5D tensors (or higher)  -- this is likely what TRT is complaining about.

**Comment by jignparm** [comment]
Closing this, since the converted  Onnx model seems fine, and the errors are due to TensorRT

**Comment by zhangyingbit** [comment]
I faced this problem too when I converted Yolact model from Pytorch->onnx->TensorRT.  I fixed it by adding the line below to the progress Pytorch->onnx.

```
torch.set_default_tensor_type('torch.FloatTensor')
torch.set_default_tensor_type('torch.cuda.FloatTensor')
```
Hope it can help.

**Comment by filipemeneses** [comment]
Had this issue, upgrading `onnxruntime` to `1.17.3` worked for me
```sh
pip install onnxruntime==1.17.3
```
